File: node/src/pagination.ts
/**
 * Pagination helpers — async iterators over page- and offset-based endpoints.
 *
 * The Legalize API uses two pagination styles:
 *
 *   - page + perPage  — laws list, webhook deliveries
 *   - limit + offset  — reforms
 *
 * Each paginated response carries `total` (true match count). That lets
 * iterators terminate without inferring end-of-stream.
 *
 * The iterators here are transport-agnostic: they delegate fetching to
 * a caller-supplied async callback.
 */

export const PAGE_MAX = 100;

/** Async iterator over a page-based endpoint (page + perPage). */
export class PageIterator<T> implements AsyncIterableIterator<T> {
  private readonly fetchPage: (page: number, perPage: number) => Promise<[T[], number]>;
  private readonly perPage: number;
  private readonly limit: number | undefined;
  private page: number;
  private yielded: number;
  private buffer: T[];
  private bufferIdx: number;
  private total: number;
  private shortPage: boolean;

  constructor(
    fetchPage: (page: number, perPage: number) => Promise<[T[], number]>,
    options: { perPage?: number; limit?: number; startPage?: number } = {},
  ) {
    const perPage = options.perPage ?? PAGE_MAX;
    if (perPage < 1 || perPage > PAGE_MAX) {
      throw new RangeError(`perPage must be between 1 and ${PAGE_MAX}`);
    }
    if (options.limit !== undefined && options.limit < 0) {
      throw new RangeError("limit must be >= 0");
    }
    this.fetchPage = fetchPage;
    this.perPage = perPage;
    this.limit = options.limit;
    this.page = options.startPage ?? 1;
    this.yielded = 0;
    this.buffer = [];
    this.bufferIdx = 0;
    this.total = 0;
    this.shortPage = false;
  }

  [Symbol.asyncIterator](): AsyncIterableIterator<T> {
    return this;
  }

  async next(): Promise<IteratorResult<T>> {
    // Serve from the in-memory buffer first.
    while (true) {
      if (this.limit !== undefined && this.yielded >= this.limit) {
        return { value: undefined, done: true };
      }

      if (this.bufferIdx < this.buffer.length) {
        const item = this.buffer[this.bufferIdx]!;
        this.bufferIdx += 1;
        this.yielded += 1;
        return { value: item, done: false };
      }

      // Buffer exhausted. Decide whether to fetch the next page.
      if (this.yielded > 0 && this.yielded >= this.total) {
        return { value: undefined, done: true };
      }
      if (this.shortPage) {
        return { value: undefined, done: true };
      }

      const [items, total] = await this.fetchPage(this.page, this.perPage);
      if (items.length === 0) {
        return { value: undefined, done: true };
      }
      this.buffer = items;
      this.bufferIdx = 0;
      this.total = total;
      this.shortPage = items.length < this.perPage;
      this.page += 1;
    }
  }
}

/** Async iterator over a limit/offset-based endpoint (reforms). */
export class OffsetIterator<T> implements AsyncIterableIterator<T> {
  private readonly fetchPage: (limit: number, offset: number) => Promise<[T[], number]>;
  private readonly batch: number;
  private readonly limit: number | undefined;
  private offset: number;
  private yielded: number;
  private buffer: T[];
  private bufferIdx: number;
  private done: boolean;

  constructor(
    fetchPage: (limit: number, offset: number) => Promise<[T[], number]>,
    options: { batch?: number; limit?: number; startOffset?: number } = {},
  ) {
    const batch = options.batch ?? 100;
    if (batch < 1) {
      throw new RangeError("batch must be >= 1");
    }
    if (options.limit !== undefined && options.limit < 0) {
      throw new RangeError("limit must be >= 0");
    }
    this.fetchPage = fetchPage;
    this.batch = batch;
    this.limit = options.limit;
    this.offset = options.startOffset ?? 0;
    this.yielded = 0;
    this.buffer = [];
    this.bufferIdx = 0;
    this.done = false;
  }

  [Symbol.asyncIterator](): AsyncIterableIterator<T> {
    return this;
  }

  async next(): Promise<IteratorResult<T>> {
    while (true) {
      if (this.limit !== undefined && this.yielded >= this.limit) {
        return { value: undefined, done: true };
      }

      // Serve from the current batch buffer first.
      if (this.bufferIdx < this.buffer.length) {
        const item = this.buffer[this.bufferIdx]!;
        this.bufferIdx += 1;
        this.yielded += 1;
        return { value: item, done: false };
      }

      // Buffer drained. If a prior batch told us we were done, stop.
      if (this.done) return { value: undefined, done: true };

      const [items, total] = await this.fetchPage(this.batch, this.offset);
      if (items.length === 0) {
        this.done = true;
        return { value: undefined, done: true };
      }
      this.buffer = items;
      this.bufferIdx = 0;
      this.offset += items.length;
      // Flag termination for the next round-trip; still serve this batch.
      if (this.offset >= total) this.done = true;
      else if (items.length < this.batch) this.done = true;
    }
  }
}

File: go/pagination.go
package legalize

import "context"

// PageMax is the server-enforced upper bound on per-page sizes.
const PageMax = 100

// LawsIter lazily iterates over every law that matches a listing or
// search. Call Next until it returns false. After the loop ends, Err
// reports any error encountered mid-stream.
//
// The iterator fetches pages on demand; there are no goroutines or
// buffers to drain, so early exit is cost-free.
type LawsIter struct {
	fetch   func(ctx context.Context, page, perPage int) ([]LawSearchResult, int, error)
	perPage int
	limit   int
	page    int
	// In-progress page buffer + cursor.
	buf       []LawSearchResult
	idx       int
	total     int
	yielded   int
	done      bool
	lastErr   error
	started   bool
	emptyPage bool
}

// NewLawsIter builds an iterator. Exposed for resource helpers;
// callers should reach for LawsService.Iter.
func newLawsIter(
	fetch func(ctx context.Context, page, perPage int) ([]LawSearchResult, int, error),
	perPage int,
	limit int,
) *LawsIter {
	if perPage < 1 || perPage > PageMax {
		perPage = PageMax
	}
	if limit < 0 {
		limit = 0
	}
	return &LawsIter{
		fetch:   fetch,
		perPage: perPage,
		limit:   limit,
		page:    1,
	}
}

// Next advances the iterator by one item. Returns (item, true, nil)
// until exhausted; (zero, false, nil) on normal end; (zero, false,
// err) on any mid-stream error. Safe to call in a for loop:
//
//	for {
//	    law, ok, err := it.Next(ctx)
//	    if err != nil { return err }
//	    if !ok { break }
//	    // ...
//	}
func (it *LawsIter) Next(ctx context.Context) (LawSearchResult, bool, error) {
	if it.done {
		return LawSearchResult{}, false, it.lastErr
	}
	if it.limit > 0 && it.yielded >= it.limit {
		it.done = true
		return LawSearchResult{}, false, nil
	}
	if it.idx >= len(it.buf) {
		if it.emptyPage || (it.started && it.yielded >= it.total && it.total > 0) {
			it.done = true
			return LawSearchResult{}, false, nil
		}
		if it.started && len(it.buf) < it.perPage && len(it.buf) > 0 {
			// Prior page was short; last page already drained.
			it.done = true
			return LawSearchResult{}, false, nil
		}
		items, total, err := it.fetch(ctx, it.page, it.perPage)
		if err != nil {
			it.done = true
			it.lastErr = err
			return LawSearchResult{}, false, err
		}
		it.started = true
		it.buf = items
		it.idx = 0
		it.total = total
		if len(items) == 0 {
			it.emptyPage = true
			it.done = true
			return LawSearchResult{}, false, nil
		}
		it.page++
	}
	item := it.buf[it.idx]
	it.idx++
	it.yielded++
	return item, true, nil
}

// Err returns the last error encountered, if any.
func (it *LawsIter) Err() error { return it.lastErr }

// ReformsIter lazily iterates over reforms for a law. Uses the
// offset/limit pagination the reforms endpoint exposes.
type ReformsIter struct {
	fetch   func(ctx context.Context, batch, offset int) ([]Reform, int, error)
	batch   int
	limit   int
	offset  int
	buf     []Reform
	idx     int
	total   int
	yielded int
	done    bool
	lastErr error
	started bool
}

func newReformsIter(
	fetch func(ctx context.Context, batch, offset int) ([]Reform, int, error),
	batch int,
	limit int,
) *ReformsIter {
	if batch < 1 {
		batch = 100
	}
	if limit < 0 {
		limit = 0
	}
	return &ReformsIter{fetch: fetch, batch: batch, limit: limit}
}

// Next advances the reforms iterator by one item.
func (it *ReformsIter) Next(ctx context.Context) (Reform, bool, error) {
	if it.done {
		return Reform{}, false, it.lastErr
	}
	if it.limit > 0 && it.yielded >= it.limit {
		it.done = true
		return Reform{}, false, nil
	}
	if it.idx >= len(it.buf) {
		if it.started {
			it.offset += len(it.buf)
			if it.offset >= it.total && it.total > 0 {
				it.done = true
				return Reform{}, false, nil
			}
			if len(it.buf) < it.batch && len(it.buf) > 0 {
				it.done = true
				return Reform{}, false, nil
			}
		}
		items, total, err := it.fetch(ctx, it.batch, it.offset)
		if err != nil {
			it.done = true
			it.lastErr = err
			return Reform{}, false, err
		}
		it.started = true
		it.buf = items
		it.idx = 0
		it.total = total
		if len(items) == 0 {
			it.done = true
			return Reform{}, false, nil
		}
	}
	item := it.buf[it.idx]
	it.idx++
	it.yielded++
	return item, true, nil
}

// Err returns the last error encountered.
func (it *ReformsIter) Err() error { return it.lastErr }
